### DISCUSSION TITLE
Rescue data from un-importable pool

### DISCUSSION THREAD

**Discussion by derekdai** [openzfs/zfs#14168] [Q&A]
I have a single-disk raidz pool. When I just learned to use zfs many years ago, I mistakenly added a file to it as stripe vdev, then I realize that I couldn't remove it from the pool, then I panicked,  delete it before data writing into the file, and then reboot system, as a result, the pool can no longer be imported.

This pool contains personal data that I don't want to lose, and finally saved as a disk image with `dd`. I would like to ask if there is any way to get the data out of it?

`zpool` can recognize the image as a zpool
```shell
# zpool import -F -X -d .
   pool: data-pool
     id: 3292954182846563831
  state: UNAVAIL
status: The pool was last accessed by another system.
 action: The pool cannot be imported due to damaged devices or data.
   see: https://openzfs.github.io/openzfs-docs/msg/ZFS-8000-EY
 config:

        data-pool                      UNAVAIL  insufficient replicas
          personal-data.zfs  UNAVAIL
```

**Comment by rincebrain** [comment]
Try setting `zfs_max_missing_tvds` to 1 and doing `zpool import -o readonly=on`?

**Reply by derekdai** [reply]
Hi, I also googled your suggestion to learn about `zfs_max_missing_tvds` parameter, thanks for the information, but in my case, `zpool` still reports
```
cannot import 'data-pool': one or more devices is currently unavailable
```

My commands
```shell
# echo 3 >> /sys/module/zfs/parameters/zfs_max_missing_tvd
# cat /sys/module/zfs/parameters/zfs_max_missing_tvds
3
# zpool import -d . -o readonly=on -m -N -f -F -X data-pool
cannot import 'data-pool': one or more devices is currently unavailable
```

**Reply by derekdai** [reply]
```
# zdb -l personal-data.zfs 
------------------------------------
LABEL 0
------------------------------------
    version: 5000
    name: 'data-pool'
    state: 0
    txg: 84811
    pool_guid: 3292954182846563831
    errata: 0
    hostid: 16845321
    hostname: 'xxxx'
    top_guid: 9059809511536026621
    guid: 9059809511536026621
    vdev_children: 3
    vdev_tree:
        type: 'disk'
        id: 0
        guid: 9059809511536026621
        path: '/dev/sdb3'
        whole_disk: 0
        metaslab_array: 33
        metaslab_shift: 35
        ashift: 12
        asize: 3998781800448
        is_log: 0
        DTL: 93
        create_txg: 4
    features_for_read:
    labels = 0 1 
failed to unpack label 2
failed to unpack label 3
```

About these two lines
```
failed to unpack label 2
failed to unpack label 3
```

Does it means the pool actually lost two top level vdevs?

**Reply by rincebrain** [reply]
No.

vdevs have four labels, two near the front of the disk, two near the rear.

If it can't find labels 2 and 3, that means either the device was truncated, or otherwise is not the size ZFS expects it to be.

**Reply by derekdai** [reply]
Got it.

Is there any other tool I can try to recover the data?
